File: 0670 Maximum Swap/0670.cpp
class Solution {
public:
    int maximumSwap(int num) {
        string s = to_string(num);
        int n = s.size();
        vector<int> last(10, -1);

        for (int i = 0; i < n; i++) last[s[i] - '0'] = i;

        for (int i = 0; i < n; i++){
            for (int x = 9; x > s[i] - '0'; x--){
                int j = last[x];
                if (j > i){
                    swap(s[i], s[j]);
                    return stoi(s);
                }
            }
        }

        return num;
    }
};

File: 0670 Maximum Swap/0670.py
class Solution(object):
    def maximumSwap(self, num):
        s = str(num)
        last = {x: i for i, x in enumerate(s)}

        for i, c in enumerate(s):
            for x in range(9, int(c), -1):
                j = last.get(str(x), -1)
                if j > i:
                    ans = list(s)
                    ans[i], ans[j] = ans[j], ans[i]
                    return int("".join(ans))
        
        return num
        
